clients/s3/s3.go

```diff
@@ -0,0 +1,235 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package s3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	cloudscale "github.com/cloudscale-ch/cloudscale-go-sdk"
+)
+
+// S3EndpointFormat is the endpoint for the S3 API without the region
+const S3EndpointFormat = "https://objects.%s.cloudscale.ch"
+
+// IsErrorNotFound helper function to test for BucketNotFound error
+func IsErrorNotFound(err error) bool {
+	if errResp, ok := err.(*cloudscale.ErrorResponse); ok {
+		return errResp.StatusCode == 404
+	} else if awsErr, ok := err.(awserr.Error); ok {
+		code := awsErr.Code()
+		return code == s3.ErrCodeNoSuchBucket || code == "NotFound"
+	}
+	return false
+}
+
+// Service defines S3 Client operations
+type Service interface {
+	CreateOrUpdateBucket(ctx context.Context, userID, bucketName, region string, cannedACL *string, tags *map[string]string) (*cloudscale.ObjectsUser, error)
+	GetBucketInfo(ctx context.Context, userID, bucketName, region string) (*cloudscale.ObjectsUser, error)
+	DeleteBucket(ctx context.Context, userID, bucketName, region string) error
+}
+
+// Client implements S3 Client
+type Client struct {
+	cloudscaleClient *cloudscale.Client
+}
+
+// NewClient creates a new S3 Client with provided Cloudscale credentials
+func NewClient(ctx context.Context, cloudscaleToken string, httpClient *http.Client) Service {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+	c := &Client{
+		cloudscaleClient: cloudscale.NewClient(httpClient),
+	}
+	c.cloudscaleClient.AuthToken = cloudscaleToken
+
+	return c
+}
+
+// CreateOrUpdateBucket creates or updates the supplied S3 bucket with provided
+// specification
+func (c *Client) CreateOrUpdateBucket(ctx context.Context, userID, bucketName, region string, cannedACL *string, tags *map[string]string) (*cloudscale.ObjectsUser, error) {
+	bucketTags := map[string]string{}
+	if tags != nil {
+		bucketTags = *tags
+	}
+	objectUserRequest := &cloudscale.ObjectsUserRequest{
+		DisplayName: bucketName,
+		Tags:        bucketTags,
+	}
+	var objectUser *cloudscale.ObjectsUser
+	existingUser, err := c.getExistingBucketUser(ctx, userID, bucketName, region)
+	if IsErrorNotFound(err) {
+		objectUser, err = c.cloudscaleClient.ObjectsUsers.Create(ctx, objectUserRequest)
+		if err != nil {
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
+	} else {
+		err := c.cloudscaleClient.ObjectsUsers.Update(ctx, existingUser.ID, objectUserRequest)
+		if err != nil {
+			return nil, err
+		}
+		objectUser, err = c.getExistingBucketUser(ctx, userID, bucketName, region)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Totally unrelated to Crossplane, but this if/else block might be easier to follow as a switch statement.